### PR TITLE
feat: created pv for core app

### PR DESCRIPTION
**Pull request by Oleksandr6676**
Created pv, pvc and set mounting to core app pod

### CODE REVIEW

**Comment by Oleksandr6676** [comment]
Hi Jan, I wasn't sure where to mount data inside the container, memory.py?

**Comment by Oleksandr6676** [comment]
It seems like I broke  our tests due to incorrect commit message in the first commit. I should have used amend command instead of pushing one more commit...

**Comment by JanMate** [comment]
> Hi Jan, I wasn't sure where to mount data inside the container, memory.py?

Nope, python files contains only python code - the logic. For storing data, we are gonna use JSON file and parse it with python. Memory usually means non-persistent data storage. Probably, we should rename memory.py file and the path of the mounted file set to something like '/data/'. For money transfer we needs 'users.json' and 'transactions.json' files. So set them to mounting.

Anyway, great catch! 👍🏽 🚀

**Comment by JanMate** [comment]
> It seems like I broke our tests due to incorrect commit message in the first commit. I should have used amend command instead of pushing one more commit... I'll try to fix it

Sure, your commits must respects [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) due to the semantic versioning.

I recommend you to squash first two commits to avoid the revert commit and set the commit message to adhere the conventions.

File: kubernetes/core-pod.yml [hunk 1]
```diff
@@ -23,3 +23,10 @@ spec:
           port: 5000
         initialDelaySeconds: 10
         timeoutSeconds: 40
+      volumeMounts:
+        - name: core-volume
+          mountPath: /core/data/memory.py
```
Comment (JanMate) [inline kubernetes/core-pod.yml:28]: Please, don't mount a code, config and source data files together. It's prefer to place them into separate folders.

File: kubernetes/storage.yml [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+kind: PersistentVolumeClaim
+metadata: 
+  name: core-pvc
+spec: 
+  storageClassName: hostpath
+  accessModes: 
+    - ReadWriteOnce
+  resources:
+    requests:
+      storage: 1Gi
```
Comment (JanMate) [inline kubernetes/storage.yml:26]: YAML files should contain new line on the end of the file.

**Review by JanMate — CHANGES_REQUESTED** [review]
Please, fix the mounting and commit messages.

**Comment by JanMate** [comment]
> > It seems like I broke our tests due to incorrect commit message in the first commit. I should have used amend command instead of pushing one more commit... I'll try to fix it
> 
> Sure, your commits must respects [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) due to the semantic versioning.
> 
> I recommend you to squash first two commits to avoid the revert commit and set the commit message to adhere the conventions.

I have already done it because I needed to test semantic release.

File: kubernetes/core-pod.yml [hunk 3]
```diff
@@ -23,3 +23,12 @@ spec:
           port: 5000
         initialDelaySeconds: 10
         timeoutSeconds: 40
+      volumeMounts:
+        - name: core-volume
+          mountPath: /app-data/transactions.json
+        - name: core-volume
+          mountPath: /app-data/users.json
```
Comment (JanMate) [inline kubernetes/core-pod.yml:30]: So sorry for scam. I didn't realize the mountPath param means directory path, not the certain file.

```suggestion
        - name: core-volume
          mountPath: /data/
```

**Review by JanMate — CHANGES_REQUESTED** [review]
Anyway, will it work with the core application if we implement the store logic? Please, verify and discuss if you see any issus with it.

**Comment by JanMate** [comment]
Regarding filesystem group permissions check this out: https://stackoverflow.com/questions/43544370/kubernetes-how-to-set-volumemount-user-group-and-file-permissions.

**Review by JanMate — CHANGES_REQUESTED** [review]
As we discussed today, it requires to set file system group permission to mounted volume.

**Comment by Oleksandr6676** [comment]
@JanMate I updated docker image with uid and gid and pushed updated image to my own repository on dockerhub  for testing purposes. https://hub.docker.com/repository/docker/oleksandr6676/flask_image
I also set  fsGroup in core-pod.yml to gid value but it still doesn't work. I'm facing the same issue with permission denied.
I found some issues on stackoverflow  and github which says that pv with type hostPath does not support security context.
https://stackoverflow.com/questions/66025328/kubernetes-fsgroup-not-changing-file-ownership-on-persistentvolume
https://github.com/kubernetes/minikube/issues/1990

**Comment by Oleksandr6676** [comment]
> @JanMate I updated docker image with uid and gid and pushed updated image to my own repository on dockerhub for testing purposes. https://hub.docker.com/repository/docker/oleksandr6676/flask_image I also set fsGroup in core-pod.yml to gid value but it still doesn't work. I'm facing the same issue with permission denied. I found some issues on stackoverflow and github which says that pv with type hostPath does not support security context. https://stackoverflow.com/questions/66025328/kubernetes-fsgroup-not-changing-file-ownership-on-persistentvolume [kubernetes/minikube#1990](https://github.com/kubernetes/minikube/issues/1990)

Shall we use another pv type then?

**Comment by JanMate** [comment]
> > @JanMate I updated docker image with uid and gid and pushed updated image to my own repository on dockerhub for testing purposes. https://hub.docker.com/repository/docker/oleksandr6676/flask_image I also set fsGroup in core-pod.yml to gid value but it still doesn't work. I'm facing the same issue with permission denied. I found some issues on stackoverflow and github which says that pv with type hostPath does not support security context. https://stackoverflow.com/questions/66025328/kubernetes-fsgroup-not-changing-file-ownership-on-persistentvolume [kubernetes/minikube#1990](https://github.com/kubernetes/minikube/issues/1990)
> 
> Shall we use another pv type then?

We should because this stands in the [k8s docs](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#types-of-persistent-volumes):
> hostPath - HostPath volume (for single node testing only; WILL NOT WORK in a multi-node cluster; consider using local volume instead)

So consider 'local' or 'ntfs' then.

**Comment by Oleksandr6676** [comment]
Tested with updated PV, Dockerfile and securityContext. Issue with permission seems to be fixed

**Comment by Oleksandr6676** [comment]
> Tested with updated PV, Dockerfile and securityContext. Issue with permission seems to be fixed

@JanMate Noticed 1 more interesting thing. It's not connected with PV itself but I'm now getting the same 'permission denied' error inside the 'core' folder. Maybe it's because this folder was copied to the docker image and not created by 'core' user?

**Comment by Oleksandr6676** [comment]
> > Tested with updated PV, Dockerfile and securityContext. Issue with permission seems to be fixed
> 
> @JanMate Noticed 1 more interesting thing. It's not connected with PV itself but I'm now getting the same 'permission denied' error inside the 'core' folder. Maybe it's because this folder was copied to the docker image and not created by 'core' user?

I fixed "permission denied" error in this PR/commit:
https://github.com/JanMate/happy_bank_core/pull/30/commits/e9327fdad612e7a06bd598c621a570e498a363ba
Core folder is now accessible.

**Review by JanMate — APPROVED** [review]

**Pull request merged**